tests/gh-api-helpers.js

```diff
@@ -0,0 +1,191 @@
+
+const octokit = new Octokit({
+  auth: process.env.GITHUB_TOKEN,
+})
+
+const helpers = {
+  scheduleAndRun: async (scheduler_res_path, runName) => {
+    return await octokit.request(
+      'POST /repos/{owner}/{repo}/actions/workflows/{workflow_id}/dispatches',
+      {
+        owner,
+        repo,
+        workflow_id,
+        ref,
+        inputs: {
+          scheduler_res_path,
+          run_name: runName,
+        },
+        headers: {
+          'X-GitHub-Api-Version': '2022-11-28',
+        },
+      },
+    )
+  },
+  rerunFailed: async (runId) => {
+    return await octokit.request(
+      'POST /repos/{owner}/{repo}/actions/runs/{run_id}/rerun-failed-jobs',
+      {
+        owner,
+        repo,
+        run_id: runId,
+        headers: {
+          'X-GitHub-Api-Version': '2022-11-28',
+        },
+      },
+    )
+  },
+  getWorkflowRun: async (runName) => {
+    const pastFiveMinutes = new Date(Date.now() - 5 * 60 * 1000)
+      .toISOString()
+      .slice(0, 19)
+    const retryTimeout = 30000
+    const retryInterval = 1000
+
+    for (let retry = 0; retry < retryTimeout / retryInterval; retry++) {
+      await new Promise((resolve) => setTimeout(resolve, retryInterval))
+      const workflows = await octokit.request(
+        'GET /repos/{owner}/{repo}/actions/workflows/{workflow_id}/runs{?created}',
+        {
+          owner,
+          repo,
+          workflow_id,
+          created: `>=${pastFiveMinutes}`,
+          headers: {
+            'X-GitHub-Api-Version': '2022-11-28',
+          },
+        },
+      )
+
+      const workflow_runs = workflows.data.workflow_runs
+      if (workflow_runs.length === 0) {
+        continue
+      }
+
+      // Using this hack because `dispatches` does not return the run id
+      const workflow = workflow_runs.find((run) => run.name === runName)
+      if (workflow) {
+        return workflow
+      }
+    }
+  },
+  getJobs: async (runId) => {
+    const jobsRes = await octokit.request(
+      'GET /repos/{owner}/{repo}/actions/runs/{run_id}/jobs',
+      {
+        owner,
+        repo,
+        run_id: runId,
+        headers: {
+          'X-GitHub-Api-Version': '2022-11-28',
+        },
+      },
+    )
+
+    return jobsRes.data.jobs
+  },
+  // this is ugly, should we consider using webhooks instead?
+  waitForRunCompletion: async (runId) => {
+    console.log(`Waiting for workflow run ${runId} to complete...`)
+    const pollingTimeout = 300000
+    const retryInterval = 5000
+
+    for (
+      let pollingCounter = 0;
+      pollingCounter < pollingTimeout / retryInterval;
+      pollingCounter++
+    ) {
+      await new Promise((resolve) => setTimeout(resolve, retryInterval))
+
+      const workflow = await octokit.request(
+        'GET /repos/{owner}/{repo}/actions/runs/{run_id}',
+        {
+          owner,
+          repo,
+          run_id: runId,
+          headers: {
+            'X-GitHub-Api-Version': '2022-11-28',
+          },
```

Comment:
I find these repetitions a bit iffy.  Is there a reason why the owner/repo/run_id need to be passed as path template parameters _and_ in the parameter object?  I'd prefer
```
const workflow = await octokit.request(`GET /repos/${owner}/${repo}/actions/runs/${runId}`)
```
but then I'm wondering whether octokit does not offer a higher-level API - It seems that here, all it adds is the auth header.
With respect to the `X-GitHub-Api-Version` header:  If it is really required, then it would be better to have axios automatically add it instead of repeating it in every API invocation.

Reply:
looks like we can do really nice things about this using: https://github.com/octokit/request.js/?tab=readme-ov-file#requestdefaults - thanks for pointing out